Added and configured cspell plugin to eslint

devcer opened the pull request:
## :bookmark_tabs: Summary

Added and configured cspell plugin to eslint

Resolves #3601 
Resolves #3642 (alternative PR for the same feature)

## :straight_ruler: Design Decisions

Describe the way your implementation works or what design decisions you made if applicable.
* Installed cspell package to package.json
* Configured the plugin in eslintrc.json
* Ran `pnpm lint` to verify its working.

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [x] :computer: have added unit/e2e tests (if appropriate)
- [x] :bookmark: targeted `develop` branch

sidharthv96 commented:
Hi @devcer, thanks for the PR. Working fine.

There are 1358 problems detected in ESLint, out of which 1263 is related to `@cspell`, which totally obliterates any actual warning we need to see.

There are 2 things we should do.
1. Integrate the cspell.json added in https://github.com/mermaid-js/mermaid/pull/3600.
2. Fix new the errors like @SeanKilleen did in #3600 so that this PR doesn't introduce any new warnings.

SeanKilleen commented:
Hopefully by merging in the develop branch, those things will be resolved. 👍

aloisklink commented:
> Hopefully by merging in the develop branch, those things will be resolved

I believe this branch is already up to date with `develop`.

The issue is that all of the new warnings are in JavaScript/TypeScript files, which PR #3600 didn't cover (it only checked documentation markdown files).

I tried modifying the spellchecker ESLint rule to ignore identifiers (e.g. variable names), and `"strings"` and `` `strings` `` (basically all that's leftover are comments), and that seems to have cut down the warnings to only 323, and quite a lot look like valid spelling mistakes.

```js
   "@cspell/spellchecker": ["warn", {
      // too many false positives for variable names
      "checkIdentifiers": false,
      "checkStrings": false,
      "checkStringTemplates": false
    }],
```

devcer commented:
Thanks for the response sidharth.
> Hi @devcer, thanks for the PR. Working fine.
> 
> There are 1358 problems detected in ESLint, out of which 1263 is related to `@cspell`, which totally obliterates any actual warning we need to see.
> 
> There are 2 things we should do.
> 
> 1. Integrate the cspell.json added in [Automated docs spell-checking via GitHub Actions (and address all reported issues) #3600](https://github.com/mermaid-js/mermaid/pull/3600).
I'm looking into this.
> 2. Fix new the errors like @SeanKilleen did in [Automated docs spell-checking via GitHub Actions (and address all reported issues) #3600](https://github.com/mermaid-js/mermaid/pull/3600) so that this PR doesn't introduce any new warnings.
Since fixing things here would be a code-level change, I would not recommend fixing all the lint warnings in a single PR. I can fix for a couple of files as an example. There is a higher chance of things getting messed up or broken.

Let me know what you think.

devcer commented:
I don't think there is a way to point `cspell.json` for the eslint. I can however configure the following options listed here -> https://www.npmjs.com/package/@cspell/eslint-plugin

Let me know what do you guys think.

aloisklink commented:
> Since fixing things here would be a code-level change, I would not recommend fixing all the lint warnings in a single PR. I can fix for a couple of files as an example. There is a higher chance of things getting messed up or broken.
>
> Let me know what you think.

If you set the following settings in `.eslintrc.json` `rules` field, it should only have warnings in code comments, so there'll be very low-risk of anything getting messed up (from a quick look, code-level warnings are mostly false positives anyway, it seems to me that it only makes sense to check comments):

```js
   "@cspell/spellchecker": ["warn", {
      // too many false positives for variable names
      "checkIdentifiers": false,
      "checkStrings": false,
      "checkStringTemplates": false
    }],
``` 

Although, it will probably be a lot of work to fix all the spelling issues (30 minutes to 1 hour or more). You'll probably have to:
  - Fix spelling issue if misspelled
  - Add to `ignoreWords` if false positive (e.g. for things like `dagre`, which is the name of a library) 
  - Fix camelCase capitalization for things that should be multiple words, (e.g. `mymermaid` should be `myMermaid` for cspell to realize it's two words)

> I don't think there is a way to point cspell.json for the eslint.

My gut feeling is that `cSpell.json` is automatically loaded, but I can't find any documentation confirming that.

devcer commented:
@aloisklink I have added the rules specified to `.eslintrc.json`, configured all the ignore words from `cspell.json`, and also fixed the lint errors in a few files. these were mostly comments so I'm confident that nothing would break on the functionality side of things.

Let me know if there is anything to be done here.

aloisklink commented on myWords.txt:
```diff
@@ -0,0 +1,14 @@
+customizability
+Gantt
+jison
+knsv
+Knut
+mindmap
+Mindmaps
+mitigations
+sandboxed
+Sveidqvist
+verdana
+Visio
+vitepress
+colour
```
By the way, you don't need this file.

Just add any words to `cSpell.json`'s `words` array, and cspell eslint should automatically load it:

https://github.com/mermaid-js/mermaid/blob/9ac3992fd2e9b679d79b9e798c0918253d42f608/cSpell.json#L4-L17

devcer replied:
it was not loading automatically. so I added this file.

aloisklink replied:
It seems to works fine on my computer if I delete the `myWords.txt` file and instead add works to `cSpell.json`.

It might be worth deleting the `myWords.txt` file and checking again if it works. If it still doesn't load the words automatically from `cSpell.json`, never mind; maybe it's just a Linux/MacOS/Windows difference.

devcer replied:
You are right. I was using `--cache`  so it was not working properly. I tested it now and added some words to ignore words list. I have also pushed many eslint error fixes. There are still a few and I couldn't judge if I have to add those words to ignore words or just use eslint-disable since it occurred only once. Take a look once now.

aloisklink replied:
Awesome! Thanks for all your work, I went ahead and fixed the rest of the cspell eslint warnings, see my main comment: https://github.com/mermaid-js/mermaid/pull/3604#pullrequestreview-1143273787

aloisklink reviewed (changes requested):
> @aloisklink I have added the rules specified to .eslintrc.json, configured all the ignore words from cspell.json, and also fixed the lint errors in a few files. these were mostly comments so I'm confident that nothing would break on the functionality side of things.
> 
> Let me know if there is anything else to be done here.

Great job! There's still 148 warnings due to `@cspell/spellchecker` that we want to fix before merging. It's probably 30 minutes to an hour of slowly fixing all of the issues.

aloisklink commented on packages/mermaid/src/dagre-wrapper/intersect/index.js:
```diff
@@ -1,5 +1,6 @@
 /*
- * Borrowed with love from from dagrge-d3. Many thanks to cpettitt!
+ // eslint-disable-next-line @cspell/spellchecker
```
```suggestion
```

We should probably just add `cpettitt` to the list of ignored words.

aloisklink commented on packages/mermaid/src/diagrams/gantt/ganttDb.spec.ts:
```diff
@@ -1,3 +1,4 @@
+// eslint-disable-next-line @cspell/spellchecker
```
```suggestion
```

Please remove and just add `ts-nocheck` to the ignored words list (we probably use `ts-nocheck` a lot).

aloisklink commented on packages/mermaid/src/diagrams/sequence/sequenceRenderer.ts:
```diff
@@ -1,3 +1,4 @@
+// eslint-disable-next-line @cspell/spellchecker
```
```suggestion
```

Please remove and just add `ts-nocheck` to the ignored words list (we probably use `ts-nocheck` a lot).

aloisklink commented on packages/mermaid/src/diagrams/user-journey/journeyRenderer.ts:
```diff
@@ -1,3 +1,4 @@
+// eslint-disable-next-line @cspell/spellchecker
```
```suggestion
```

Please remove and just add `ts-nocheck` to the ignored words list (we probably use `ts-nocheck` a lot).

aloisklink commented on packages/mermaid/src/utils.ts:
```diff
@@ -1,3 +1,4 @@
+// eslint-disable-next-line @cspell/spellchecker
```
```suggestion
```

Please remove and just add `ts-nocheck` to the ignored words list (we probably use `ts-nocheck` a lot).

aloisklink reviewed (approved):
Hi @devcer, I hope you don't mind, I went ahead and tried to fix the rest of the `@cspell/eslint` issues, and I've pushed them to your branch.

Now there are no longer any `@cspell/eslint` warnings!

There's still some `// eslint-disable-next-line @cspell/spellchecker` I think we could remove by adding stuff to the ignoreWords list in `cSpell.json`, but I don't think they're blockers.

**For any reviewers testing this, you may need to delete the `.eslintcache` file to get ESLint to relint all the files.**

P.S. Thanks for also fixing some of the other ESLint warnings, like the JSDoc ones!

devcer commented:
@aloisklink Thanks for your commit. I have removed a couple of more `eslint-ignore` statements and added related words to cspell.json. I see 0 eslint warnings related to spellings.

sidharthv96 commented on .eslintrc.json:
```diff
@@ -48,6 +49,14 @@
     ],
     "json/*": ["error", "allowComments"],
+    "@cspell/spellchecker": [
+      "warn",
```
If there are 0 issues now, can't we change this to error?

devcer replied:
Done! I have made the change.
